File: impls/java/src/mal/step5_tco.java
package mal;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import mal.types.FunctionTco;
import mal.types.MalFunction;
import mal.types.MalList;
import mal.types.MalSymbol;
import mal.types.MalType;
import mal.types.RepException;

public class step5_tco {

	mal.types malTypes = new types();
	env repl_env = new env(null, null, null);
	
	public step5_tco() {
		core.ns.forEach((name, fn) -> repl_env.set(malTypes.new MalSymbol(name), fn));
	}


	public MalType eval(MalType ast, env replEnv) {
		while (true) {
			if (!(ast instanceof MalList)) {
				return eval_ast(ast, replEnv);
			} else {
				MalList inputList = (MalList) ast;
				if (inputList.items.isEmpty()) {
					return ast;
				} else {
					if ("(".equals(inputList.open)) { // actual list, not vector or map
						MalType firstUnevaluated = inputList.items.get(0);
						if (firstUnevaluated instanceof MalSymbol) {
							MalSymbol firstSymbol = (MalSymbol) firstUnevaluated;
							if (firstSymbol.name.equals("def!")) {
								MalSymbol key = (MalSymbol) inputList.items.get(1);
								MalType value = inputList.items.get(2);
								MalType valueEvaled = eval(value, replEnv);
								replEnv.set(key, valueEvaled);
								return valueEvaled;
							} else if (firstSymbol.name.equals("let*")) {
								// Create new env
								env newEnv = new env(replEnv, null, null);
								// Eval the key-value pairs in the new env
								MalList keyValues = (MalList) inputList.items.get(1);
								for (int i = 0; i < keyValues.items.size(); i += 2) {
									MalSymbol key = (MalSymbol) keyValues.items.get(i);
									MalType value = keyValues.items.get(i + 1);
									MalType valueEvaled = eval(value, newEnv);
									newEnv.set(key, valueEvaled);
								}							
								// finally, eval the body after the name-value pairs and return the result
								MalType letBody = inputList.items.get(2);
								replEnv = newEnv;
								ast = letBody;
								continue;
							} else if (firstSymbol.name.equals("do")) {
								// Evaluate all the elements of the list using eval_ast and return the final evaluated element
								for (int i = 1; i < inputList.items.size() - 1; i++) {
									eval(inputList.items.get(i), replEnv);
								}
								ast = inputList.items.get(inputList.items.size() - 1);
								continue;
							} else if (firstSymbol.name.equals("if")) {
								MalType test = eval(inputList.items.get(1), replEnv);
								if (types.MalNil.equals(test) || types.MalFalse.equals(test)) {
									if (inputList.items.size() == 3) {
										return types.MalNil;
									} else {
										ast = inputList.items.get(3);
										continue;
									}
								} else {
									ast = inputList.items.get(2);
									continue;
								}
							} else if (firstSymbol.name.equals("fn*")) {
								final env replEnvCopy = new env(replEnv, null, null); // required because Java wants replEnv to be final
								MalList binds = (MalList) inputList.items.get(1);
								MalType expr = inputList.items.get(2);
								MalFunction fn = malTypes.new MalFunction() {
									MalType apply(MalList args) {
										// Copy the lists binds and args lists around, if one of the 'binds' is '&'
										// capture every args afterwards as a list
										MalList bindsNew = malTypes.new MalList(new ArrayList<>());
										MalList argsNew = malTypes.new MalList(new ArrayList<>());
										for (int i = 0; i < binds.items.size(); i++) {
											MalSymbol bindsi = (MalSymbol) binds.items.get(i);
											MalType argsi = (i < args.items.size()) ? args.items.get(i) : types.MalNil;
											if (bindsi.name.equals("&")) {
												bindsi = (MalSymbol) binds.items.get(i + 1); // & more
												if (i < args.items.size()) {
													argsi = malTypes.new MalList(args.items.subList(i, args.items.size()));
												} else {
													argsi = malTypes.new MalList(new ArrayList<>());
												}
												bindsNew.items.add(bindsi);
												argsNew.items.add(argsi);
												break;
											} else {
												bindsNew.items.add(bindsi);
												argsNew.items.add(argsi);
											}
										}
										env newEnv = new env(replEnvCopy, bindsNew, argsNew);
										return eval(expr, newEnv);
									}
								};
								return malTypes.new FunctionTco(expr, binds, replEnv, fn);
							}
						}
						
						// regular function application
						MalList evaluated = (MalList) eval_ast(inputList, replEnv);
						MalType funcOrFunctionTco = evaluated.items.get(0);
						MalList funcArgs = malTypes.new MalList(evaluated.items.subList(1, evaluated.items.size()));
						if (funcOrFunctionTco instanceof MalFunction) {
							MalFunction func = (MalFunction) funcOrFunctionTco;
							return func.apply(funcArgs);
						} else {
							FunctionTco funcTco = (FunctionTco) funcOrFunctionTco;
							ast = funcTco.ast;
							replEnv = new env(funcTco.functionEnv, funcTco.params, funcArgs);
							continue;
						}
					} else {
						// vector and pseudo-maps literals evaluate the contents with eval_ast and wrap in square/braces
						MalList evaluated = (MalList) eval_ast(inputList, replEnv);
						MalList result = malTypes.new MalList(evaluated.items);
						result.open = inputList.open;
						result.close = inputList.close;
						return result;
					}
				}
			}
		}
	}

	public MalType eval_ast(MalType ast, env replEnv) {
		if (ast instanceof MalSymbol) {
			MalSymbol astSymbol = (MalSymbol) ast;
			return replEnv.get(astSymbol);
		} else if (ast instanceof MalList) {
			MalList args = (MalList) ast;
			List<MalType> evaluated = args.items.stream().map(arg -> eval(arg, replEnv)).collect(Collectors.toList());
			return malTypes.new MalList(evaluated);
		} else {
			return ast;
		}
	}

	public String rep(String input) {
		try {
			MalType afterRead = reader.read_str(input);
			MalType afterEval = this.eval(afterRead, repl_env);
			return printer.pr_str(afterEval, true);
		} catch (RepException rex) {
			return rex.getMessage();
		} catch (Exception e) {
			return "EOF"; // generic error
		}
	}

	public static void main(String... args) {
		step5_tco rp = new step5_tco();

		// Local Test
		// System.out.println(rp.rep("(= [(list)] (list []))"));

		// Functions defined in MAL itself
		rp.rep("(def! not (fn* (a) (if a false true)))"); // (not <expr>)

		String input;
		do {
			input = System.console().readLine("user> ");
			if (input != null) {
				String repResult = rp.rep(input);
				if (repResult != null) {
					System.out.println(repResult);
				}
			}
		} while (input != null);
	}
}

File: impls/clojure/src/mal/step5_tco.clj
(ns mal.step5_tco
  (:require [mal.core :as core]
            [mal.env :as env]
            [mal.reader :as reader]
            [mal.printer :as printer]))

(def repl_env
  (env/make-env nil (keys core/ns) (vals core/ns)))

(def READ reader/read_str)

(declare eval_ast)

(defn rebind [binds args] ;; update args if bindings include `&`
  (if (some #{'&} binds)
    (let [binds'  (remove #{'&} binds)
          [as bs] (split-at (dec (count binds')) args)
          args'   (concat as (list bs))]
      [binds' args'])
    [binds args]))

(defn EVAL [ast envi]
  (loop [ast ast, envi envi]
    (cond
      (not (seq? ast)) (eval_ast ast envi)
      (and (seq? ast) (empty? ast)) ast
      :else (let [[f & xs] ast]
              (condp = f
                'def! (let [val (EVAL (second xs) envi)]
                        (env/set envi (first xs) val)
                        val)
                'let* (let [newenv (env/make-env envi)
                            ps     (partition 2 (first xs))
                            body   (second xs)]
                        (dorun
                         (map (fn [[k v]] (env/set newenv k (EVAL v newenv))) ps))
                        (recur body newenv))
                'do   (do (eval_ast (butlast xs) envi)
                          (recur (last xs) envi))
                'if   (let [[test then else] xs]
                        (if (EVAL test envi) (recur then envi) (recur else envi)))
                'fn*  (let [[binds body] xs
                            func (fn [& args] ;; TODO validate number of args over binds
                                   (let [[bs' as'] (rebind binds args)
                                         newenv    (env/make-env envi bs' as')]
                                     (EVAL body newenv)))]
                        {:ast body :params binds :env envi :fn func :type :tco})
                ;; otherwise, attempt regular function application
                (let [[f & xs] (eval_ast ast envi)]
                  (if f
                    (cond
                      (fn? f) (apply f xs)
                      :else (let [{:keys [ast params env fn]} f
                                  [ps' xs'] (rebind params xs)
                                  newenv    (mal.env/make-env env ps' xs')]
                              (recur ast newenv)))
                    (throw (ex-info "Symbol not found" {})))))))))

(defn eval_ast [ast envi]
  (cond
    (symbol? ast) (env/get envi ast)
    (seq? ast)    (map #(EVAL % envi) ast)
    (vector? ast) (mapv #(EVAL % envi) ast)
    (map? ast)    (zipmap (keys ast) (map #(EVAL % envi) (vals ast)))
    :else ast))

(def PRINT printer/pr_str)

(defn rep [s]
  (-> (READ s)
      (EVAL repl_env)
      (PRINT true)))

(defn prompt []
  (print "user> ")
  (flush)
  (read-line))

;; MAIN LOOP

(defn -main [& args]
  ;; Define `not` in mal itself
  (rep "(def! not (fn* (a) (if a false true)))")

  (loop [input (prompt)]
    (when input
      (try
        (println (rep input))
        (catch Exception ex
          (when-let [message (ex-message ex)] ;; print message only when there's one
            (println "ERROR:" message))))
      (recur (prompt)))))

